### ISSUE TITLE
String interpolation not working in Markdown links using md_str macro

### ISSUE THREAD

**Issue by lungben** [JuliaLang/julia#38298]
Hi,

interpolation of Julia variables into Markdown links are not working correctly when using the `md""` string macro.
With the function `Markdown.parse` it works correctly.

MWE:

```
julia> using Markdown

julia> urlname = "https://github.com/JuliaLang/julia"
"https://github.com/JuliaLang/julia"

julia> md"hello $urlname" # string interpolation in Markdown text working correctly
  hello https://github.com/JuliaLang/julia

julia> md"Hello [Julia]($urlname)" # string interpolation in Markdown link is not working
  Hello Julia ($urlname)

# adding brackets and escaping does not help:

julia> md"Hello [Julia]($(urlname))"
  Hello Julia ($(urlname))

julia> md"Hello [Julia](\$\(urlname\))"
  Hello Julia (\$\(urlname\))

julia> Markdown.parse("Hello [Julia]($urlname)") # this works correctly
  Hello Julia (https://github.com/JuliaLang/julia)

julia> versioninfo()
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 6

```

Best Regards
Benjamin

**Comment by zerefwayne** [comment]
@fredrikekre I would like to work on this one.

**Comment by zerefwayne** [comment]
@fredrikekre

https://github.com/JuliaLang/julia/blob/6e88e31e84c56a88fbd35cf1de2f114fae9267e0/stdlib/Markdown/src/Markdown.jl#L34-L37

What does line 36 exactly do? I think when Markdown.parse(..) is called, this line is skipped, so there might be some problem here.

https://github.com/JuliaLang/julia/blob/0bedcdabeb21d0d244babb4a88c91ff75a15577f/stdlib/Markdown/test/runtests.jl#L208-L210

There should be a `Link` instead of just a Paragraph.

**Comment by zerefwayne** [comment]
@fredrikekre I think I have run into a big issue with string interpolation while using string macros. I've tried to reproduce a simple problem.

<pre><font color="#50FA7B"><b>julia&gt; </b></font>macro hello_str(s)
       println(&quot;jumbo&quot;, s)
       end
@hello_str (macro with 1 method)

<font color="#50FA7B"><b>julia&gt; </b></font>hello&quot;Hey&quot;
jumboHey

<font color="#50FA7B"><b>julia&gt; </b></font>a = 32
32

<font color="#50FA7B"><b>julia&gt; </b></font>hello&quot;Hey $a&quot;
jumboHey $a
</pre>

**Comment by zerefwayne** [comment]
https://github.com/JuliaLang/julia/issues/23481

^ The discussion here clearly states that Julia has decided against using Interpolation in String macros.

Please use `Markdown.parse()` instead of `md` macro as much as possible. If you really need the macro, please construct the string beforehand and then pass it to the macro.

**Comment by zerefwayne** [comment]
@fredrikekre I've noticed some issues with the test cases, please let me know if these need to be fixed.

https://github.com/JuliaLang/julia/blob/6e88e31e84c56a88fbd35cf1de2f114fae9267e0/stdlib/Markdown/test/runtests.jl#L208-L210

I think we should use `Link` here instead of `Paragraph`.

**Comment by fredrikekre** [comment]
> The discussion here clearly states that Julia has decided against using Interpolation in String macros. Please use Markdown.parse() instead of md macro as much as possible. If you really need the macro, please construct the string beforehand and then pass it to the macro.

The `md` string macro implements string interpolation though, it just misses some places. Whether it should be done inside links like this, I am not sure. I suppose it could lead to strange things if you happen to have `$` in an URL, but maybe we don't have to worry about that.

**Comment by zerefwayne** [comment]
Yes, I just found it out. `md` works fine with interpolations. However, the variable parsing via `esc` doesn't happen when a Link is being parsed. 

I think if we are allowing interpolations in this macro, interpolations should also happen inside [...](...).

**Comment by lungben** [comment]
For Pluto, the md string macro is very useful. My concrete use case is to add links relative to the notebook path, which requires using the `pwd()` function.

@zerefwayne @fredrikekre thanks for picking this up!

**Comment by zerefwayne** [comment]
@lungben  I will suggest you generate the string and then pass it. The interpolation for links might take time.

**Comment by fonsp** [comment]
> I suppose it could lead to strange things if you happen to have $ in an URL, but maybe we don't have to worry about that.

How about we only interpolate if the `$` is used as the first character? 

So `[...]($something)` interpoaltes, `[...](also$something)` does not.
